src/vchan_cmds.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2024 EPAM Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <domain.h>
+#include <vch.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/shell/shell.h>
+
+LOG_MODULE_REGISTER(vchan_cmd);
+
+#define ECHO_SERVER_THREAD_PRIO    7
+static struct k_thread echo_thread;
+static struct k_thread *pecho_thread = NULL;
+struct vch_handle handle;
+
+static K_THREAD_STACK_DEFINE(main_stack, 8192);
+static volatile int vchan_server_stop = 0;
+static char buf[10];
+#define PATH_TEMPLATE "/local/domain/%d/echo"
+char path[30];
+
+static void vchan_echo_server(void *p1, void *p2, void *p3)
+{
+	struct vch_handle *h = (struct vch_handle *)p1;
+	int size;
+	int count = 0;
+
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+	printk("Thread started\n");
+	while (!vchan_server_stop) {
+		size = vch_read(h, buf, 1);
+		if (size > 0) {
+			vch_write(h, buf, 1);
+		} else {
+			if (size < 0 && size != -ENOTCONN) {
+				printk("%d vch_read returned %d\r", count++, size);
+			}
+		}
+		k_sleep(K_MSEC(500));
+	}
+	printk("Thread stoped\n");
+	vch_close(h);
+}
+
+static int vchan_shell_start(const struct shell *sh, size_t argc, char **argv)
+{
+	int ret;
+	domid_t domid = 0;
+
+	if (pecho_thread) {
+		shell_warn(sh, "Server already started");
+		return 0;
+	}
+	if (argv[1]) {
+		domid = atoi(argv[1]);
+	}
+	sprintf(path, PATH_TEMPLATE, domid);
+	ret = vch_open(domid, path, 10, 10, &handle);
+	if (ret) {
+		shell_error(sh, "vchan open error %d", ret);
+		goto exit;
+	}
+	shell_print(sh, "Echo server openad at %s", path);
+	vchan_server_stop = 0;
+	k_thread_create(&echo_thread, main_stack, K_THREAD_STACK_SIZEOF(main_stack), vchan_echo_server,
+			(void *)&handle, NULL, NULL, ECHO_SERVER_THREAD_PRIO, 0, K_NO_WAIT);
+	pecho_thread = &echo_thread;
+
+exit:
+	return ret;
+}
+
+static int vchan_shell_stop(const struct shell *sh, size_t argc, char **argv)
+{
+	if (!pecho_thread) {
+		shell_warn(sh, "Server not started");
+		return 0;
+	}
+	vchan_server_stop = 1;
+	if (k_thread_join(pecho_thread, K_MSEC(1000))) {
+		shell_error(sh, "Killing thread");
+		k_thread_abort(pecho_thread);
+	}
+	pecho_thread = NULL;
+	return 0;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(
+	subcmd_vchan,
+	SHELL_CMD_ARG(start, NULL,
+		      "start echo server.\n"
+		      "\tUsage:\n"
+		      "\tvchan start [domid]\n",
+		      vchan_shell_start, 0, 1),
+	SHELL_CMD_ARG(stop, NULL,
+		      "stop server.\n",
+		      vchan_shell_stop, 0, 0),
+	SHELL_SUBCMD_SET_END);
+
+SHELL_CMD_ARG_REGISTER(vchan, &subcmd_vchan, "VCHAN commands", NULL, 2, 0);
```

Comment:
Commands must be described in help and your commit message

Reply:
Fixed